File: 0015-3sum/0015-3sum.cpp
class Solution {
public:
    vector<vector<int>> threeSum(vector<int>& nums) {
        sort(nums.begin(), nums.end());
        vector<vector<int>> op;

        for(int i=0;i<nums.size();i++)
        {
            if(i >0 && nums[i] == nums[i-1]) continue;

            int l = i+1;
            int r = nums.size()-1;

            while(l < r)
            {
                if(nums[l] + nums[r] + nums[i] == 0)
                {
                    op.push_back({nums[l], nums[r], nums[i]});

                    while(l < r && nums[l] == nums[l+1])
                    {
                        l++;
                    }
                    while(l < r && nums[r] == nums[r-1])
                    {
                        r--;
                    }

                    l++;
                    r--;
                }
                else if(nums[l] + nums[r] + nums[i] > 0)
                {
                    r--;
                }
                else
                {
                    l++;
                }
            }
        }

        return op;
    }
};

File: 0015-3sum/0015-3sum.py
class Solution:
    def threeSum(self, nums: list[int]) -> list[list[int]]:
        nums.sort()
        op = []

        for i in range(len(nums)):
            if i > 0 and nums[i] == nums[i-1]:
                continue

            l = i + 1
            r = len(nums) - 1 

            while l < r:
                sumi = nums[i] + nums[l] + nums[r]
                if sumi == 0:
                    op.append([nums[i],nums[l],nums[r]])
                    #skipping duplicates
                    while l < r and nums[l] == nums[l+1]:
                        l += 1
                    while l < r and nums[r] == nums[r-1]:
                        r -= 1
                    
                    l += 1
                    r -= 1
                    
                elif nums[i] + nums[l] + nums[r] > 0:
                    r -= 1
                else:
                    l += 1

        return op